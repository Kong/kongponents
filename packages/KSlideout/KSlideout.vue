<template>
  <div class="k-slideout">
    <transition name="fade">
      <div
        v-if="isVisible"
        class="panel-background"
        @click="e => handleClose(e, true)"/>
    </transition>
    <transition
      tag="div"
      name="slide">
      <div
        v-if="isVisible"
        :class="{isVisible: 'isVisible'}"
        class="panel">
        <button
          class="close-btn"
          @click="e => handleClose(e, true)">
          <KIcon
            icon="close"
            size="16"
            view-box="0 0 12 12" />
        </button>
        <div class="content">
          <KCard border-variant="noBorder">
            <template v-slot:body>
              <slot />
            </template>
          </KCard>
        </div>
      </div>
    </transition>
  </div>
</template>

<script>
import KCard from '@kongponents/kcard/KCard.vue'
import KIcon from '@kongponents/kicon/KIcon.vue'

export default {
  name: 'KSlideout',
  components: { KCard, KIcon },
  props: {
    isVisible: {
      type: Boolean,
      default: false
    }
  },

  mounted () {
    document.addEventListener('keydown', this.handleClose)
  },

  beforeDestroy () {
    document.removeEventListener('keydown', this.handleClose)
  },

  methods: {
    handleClose (e, forceClose) {
      if ((this.isVisible && e.keyCode === 27) || forceClose) {
        this.$emit('close')
      }
    }
  }
}
</script>

<style lang="scss">
.k-slideout .close-btn {
  svg path {
    transition: all 200ms ease;
  }
  &:hover svg path {
    fill-opacity: .5;
  }
}
</style>

<style lang="scss" scoped>
.k-slideout {
  --KCardPaddingY: 2rem;
  --KCardPaddingX: 3rem;

  .panel-background {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background: var(--tblack-45, hsla(0, 0%, 0%, .45));
    z-index: 9999;
  }
  .panel {
    position: fixed;
    top: 0;
    right: 0;
    height: 100vh;
    width: 100%;
    max-width: 500px;
    background-color: var(--twhite-1, hsla(100, 100%, 100%, 1));
    z-index: 9999;
    .close-btn {
      position: absolute;
      top: 1rem;
      left:  1rem;
      height: 16px;
      width: 16px;
      padding: 0;
      border: none;
      background: none;
      cursor: pointer;
      outline: inherit;
      transition: 200ms ease;
    }

    .content {
      height: 100%;
      overflow: auto;
      -ms-overflow-style: none;  // IE 10+
      scrollbar-width: none;  // Firefox
      &::-webkit-scrollbar { display: none; }
    }

  }

  .slide-enter-active { animation: slide .3s cubic-bezier(1.0, 0.5, 0.8, 1.0); }
  .slide-leave-active { animation: slide .3s ease reverse; }
  .fade-enter-active, .fade-leave-active { transition: opacity 500ms;}
  .fade-enter, .fade-leave-to { opacity: 0; }
}

@keyframes slide {
  0% { transform: translateX(100%); }
  100% { transform: translateX(0%); }
}
</style>
