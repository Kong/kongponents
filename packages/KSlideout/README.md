```js
<KSlideout 
  :is-visible="isToggled"
  @close="toggle">
  <div>
    <p>Anim officia eiusmod duis est consequat nulla tempor ad non magna Lorem ullamco nostrud amet. Occaecat voluptate dolor enim eiusmod do qui nulla pariatur enim. Et elit elit consequat do do duis enim est ullamco id sunt sunt amet eiusmod. Do minim mollit irure ea sunt officia minim sint eiusmod enim amet. Quis exercitation in ullamco quis aliqua.</p>
  </div>
</KSlideout>
```
