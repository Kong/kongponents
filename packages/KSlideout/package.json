{
  "name": "@kongponents/kslideout",
  "version": "6.32.11",
  "description": "KSlideout description here.",
  "main": "dist/KSlideout.umd.min.js",
  "componentName": "KSlideout",
  "source": "KSlideout.vue",
  "files": [
    "dist",
    "KSlideout.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kcard": "^6.32.11",
    "@kongponents/kicon": "^6.32.11",
    "@kongponents/styles": "^6.32.11"
  }
}
