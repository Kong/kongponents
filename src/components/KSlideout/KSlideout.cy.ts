import { mount } from 'cypress/vue'
import KSlideout from '@/components/KSlideout/KSlideout.vue'
import { h } from 'vue'

describe('KSlideout', () => {
  it('renders default slot', () => {
    mount(KSlideout, {
      props: {
        isVisible: true,
      },
      slots: {
        default: () => h('div', {}, [
          h('h1', {}, 'What\'s up default slot'),
          h('p', {}, 'Default slots are the easiest'),
        ]),
      },
    })

    cy.get('h1').should('be.visible')
    cy.get('p').should('be.visible')
  })

  it('renders props when passed', () => {
    const titleProp = 'Hello!'
    const closeButtonAlignmentProp = 'end'

    mount(KSlideout, {
      props: {
        isVisible: true,
        title: titleProp,
        closeButtonAlignment: closeButtonAlignmentProp,
      },
    })

    cy.getTestId('k-slideout-title').should('exist')
    cy.getTestId('k-slideout-title').should('be.visible')
  })

  it('renders title slot when providing both title prop and slot', () => {
    mount(KSlideout, {
      props: {
        isVisible: true,
        title: 'Title prop',
      },
      slots: {
        title: () => h('div', {}, [
          h('span', {}, 'Title slot'),
        ]),
      },
    })

    cy.getTestId('k-slideout-title').contains('Title slot')
  })

  it('renders cancel button on right when prop is used', () => {
    const closeButtonAlignmentProp = 'end'

    mount(KSlideout, {
      props: {
        isVisible: true,
        closeButtonAlignment: closeButtonAlignmentProp,
      },
    })

    cy.getTestId('close-button-end').should('exist')
    cy.getTestId('close-button-end').should('be.visible')
  })

  it('emits close when panel-background is clicked', () => {
    mount(KSlideout, {
      props: {
        isVisible: true,
        onClose: cy.spy().as('onCloseSpy'),
      },
    }).then(({ wrapper }) => wrapper)
      .as('vueWrapper')

    cy.get('@vueWrapper').then((wrapper: any) => wrapper.findComponent(KSlideout)
      .vm.$emit('close', true))

    cy.get('@onCloseSpy').should('have.been.called')

    cy.get('.panel-background').click({ force: true })
      .then(() => {
        cy.get('@onCloseSpy').should('have.been.called')
      })
  })

  it('emits close when esc key pressed', () => {
    mount(KSlideout, {
      props: {
        isVisible: true,
      },
    })

    cy.get('body').type('{esc}').then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'close')
    })
  })

  it('does not emit close event when persist prop is true', () => {
    mount(KSlideout, {
      props: {
        isVisible: true,
        preventCloseOnBlur: true,
      },
    }).then(({ wrapper }) => wrapper)
      .as('vueWrapper')

    cy.get('html').click(0, 0).then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('not.have.property', 'close')
    })
  })
})
