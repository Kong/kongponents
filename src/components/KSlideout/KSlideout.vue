<template>
  <div class="k-slideout">
    <transition name="fade">
      <div
        v-if="isVisible"
        :class="hasOverlay ? 'panel-background' : 'panel-background-transparent'"
      />
    </transition>
    <transition
      name="slide"
    >
      <div
        v-if="isVisible"
        ref="slideOutRef"
        class="panel"
        :class="{ 'is-visible': isVisible, 'border-styles': !hasOverlay }"
      >
        <!-- If there is title prop or header slot, display the header content -->
        <div
          v-if="title || hasTitle"
          class="k-slideout-header-content"
        >
          <slot name="header">
            <p class="k-slideout-title">
              {{ title }}
            </p>
          </slot>
        </div>

        <button
          :class="closeButtonAlignment === 'start' ? 'close-button-start' : 'close-button-end'"
          @click="(event: any) => emit('close')"
        >
          <KIcon
            :color="`var(--kui-color-text-neutral-stronger, ${KUI_COLOR_TEXT_NEUTRAL_STRONGER})`"
            icon="close"
            :size="KUI_ICON_SIZE_50"
          />
        </button>
        <div class="content">
          <KCard border-variant="noBorder">
            <template #body>
              <slot />
            </template>
          </KCard>
        </div>
      </div>
    </transition>
  </div>
</template>

<script lang="ts" setup>
import { computed, useSlots, ref, onMounted, onUnmounted } from 'vue'
import KCard from '@/components/KCard/KCard.vue'
import KIcon from '@/components/KIcon/KIcon.vue'
import useUtilities from '@/composables/useUtilities'
import { KUI_ICON_SIZE_50, KUI_COLOR_TEXT_NEUTRAL_STRONGER } from '@kong/design-tokens'
import { onClickOutside } from '@vueuse/core'

const props = defineProps({
  isVisible: {
    type: Boolean,
    default: false,
  },
  // controls close button alignment
  closeButtonAlignment: {
    type: String,
    default: 'start',
    validator: (value: string): boolean => {
      return ['start', 'end'].includes(value)
    },
  },
  // enable/disable overlay to be able to interact with the background content while the slideout is expanded
  hasOverlay: {
    type: Boolean,
    default: true,
  },
  // allows a host app to define the offset from the top of the page
  offsetTop: {
    type: String,
    default: '0',
  },
  title: {
    type: String,
    default: '',
  },
})

const emit = defineEmits<{
  (e: 'close'): void
}>()

const slots = useSlots()
const hasTitle = computed((): boolean => !!slots.title)
const { getSizeFromString } = useUtilities()
const slideOutRef = ref(null)

onClickOutside(
  slideOutRef,
  (event) => {
    if (event.isTrusted) {
      emit('close')
    }
  },
)

const handleClose = (e: any, forceClose = false): void => {
  if ((props.isVisible && e.keyCode === 27) || forceClose) {
    emit('close')
  }
}

onMounted(() => {
  document.addEventListener('keydown', handleClose)
})

onUnmounted(() => {
  document.removeEventListener('keydown', handleClose)
})

const offsetTopValue = computed((): string => getSizeFromString(props.offsetTop))

</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/tmp-variables';
@import '@/styles/functions';

.k-slideout {
  --KCardPaddingY: var(--kui-space-90, #{$kui-space-90});
  --KCardPaddingX: var(--kui-space-110, #{$kui-space-110});

  .k-slideout-header-content {
    display: flex;
    .k-slideout-title {
      color: var(--black-400, var(--kui-color-text-neutral, $kui-color-text-neutral));
      flex:1;
      font-size: var(--kui-font-size-40, $kui-font-size-40);
      font-weight: var(--kui-font-weight-medium, $kui-font-weight-medium);
      line-height: var(--kui-line-height-40, $kui-line-height-40);
      margin-left: var(--kui-space-50, $kui-space-50);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }
  }
  .panel {
    background-color: var(--white, var(--kui-color-background, $kui-color-background));
    display: flex;
    flex-direction: column;
    height: 100vh;
    max-width: 500px;
    overflow-y: auto;
    position: fixed;
    right: 0;
    top: v-bind('offsetTopValue');
    width: 100%;
    z-index: 9999;

    .close-button-start {
      align-self: flex-start;
      background: none;
      border: none;
      cursor: pointer;
      display: flex;
      height: auto;
      margin-left: var(--kui-space-50, $kui-space-50);
      margin-top: var(--kui-space-60, $kui-space-60);
      outline: inherit;
      position: absolute;
      transition: $tmp-animation-timing-2 ease;

      &:focus{
        box-shadow: 0 0 0 2px var(--KButtonOutlineBorder, var(--blue-500, var(--kui-color-border-primary, $kui-color-border-primary)))
      }
    }

    .close-button-end {
      align-self: flex-end;
      background: none;
      border: none;
      cursor: pointer;
      display: flex;
      height: auto;
      margin-right: var(--kui-space-50, $kui-space-50);
      margin-top: var(--kui-space-60, $kui-space-60);
      outline: inherit;
      position: absolute;
      transition: $tmp-animation-timing-2 ease;

      &:focus{
        box-shadow: 0 0 0 2px var(--KButtonOutlineBorder, var(--blue-500, var(--kui-color-border-primary, $kui-color-border-primary)))
      }
    }

    .close-button-with-title {
      margin-left: var(--kui-space-60, $kui-space-60);
      margin-right: var(--kui-space-60, $kui-space-60);
      margin-top: var(--kui-space-30, $kui-space-30);
    }

    .content {
      height: 100%;
      -ms-overflow-style: none;  // IE 10+
      scrollbar-width: none;  // Firefox
      &::-webkit-scrollbar { display: none; }
    }
  }
}

  .panel-background {
    background: var(--black-45, $tmp-color-black-45);
    bottom: 0;
    left: 0;
    position: fixed;
    right: 0;
    top: v-bind('offsetTopValue');
    z-index: 9999;
  }

  .panel-background-transparent {
    background: transparent;
    bottom: 0;
    left: 0;
    position: fixed;
    right: 0;
    top: v-bind('offsetTopValue');
    z-index: -1;
  }

  .border-styles {
    border-left: var(--kui-border-width-10, $kui-border-width-10) solid var(--grey-300, var(--kui-color-border-neutral-weak, $kui-color-border-neutral-weak));
    box-shadow: -2px 0px 5px var(--black-5, $tmp-color-black-5);
  }
</style>

<style lang="scss">
// @keyframes animations need to be un-scoped
.k-slideout {
  @keyframes slide {
    0% { transform: translateX(100%); }
    100% { transform: translateX(0%); }
  }

  .slide-enter-active { animation: slide .3s cubic-bezier(1.0, 0.5, 0.8, 1.0); }

  .slide-leave-active { animation: slide .3s ease reverse; }

  .fade-enter-active, .fade-leave-active { transition: opacity 500ms;}

  .fade-enter, .fade-leave-to { opacity: 0; }
}
</style>
