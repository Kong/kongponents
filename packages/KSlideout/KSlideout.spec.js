import { mount } from '@vue/test-utils'
import KSlideout from '@/KSlideout/KSlideout'

describe('KSlideout', () => {
  it('renders default slot', () => {
    const wrapper = mount(KSlideout, {
      propsData: {
        isVisible: true
      },
      slots: {
        default: `<div>
          <h1>Sup default</h1>
          <p>default slots are the best</p>
        </div>`
      }
    })

    const h1 = wrapper.find('h1')
    const p = wrapper.find('p')

    expect(h1.isVisible()).toBe(true)
    expect(p.isVisible()).toBe(true)
  })

  it('emits close when panel-background is clicked', () => {
    const wrapper = mount(KSlideout, {
      propsData: {
        isVisible: true
      }
    })

    const backdrop = wrapper.find('.panel-background')

    backdrop.trigger('click')
    expect(wrapper.emitted().close).toHaveLength(1)
  })

  it('emits close when esc key pressed', () => {
    const wrapper = mount(KSlideout, {
      propsData: {
        isVisible: true
      },
      attachToDocument: true
    })

    wrapper.trigger('keydown.esc')
    expect(wrapper.emitted().close).toHaveLength(1)
  })

  it('matches snapshot', () => {
    const wrapper = mount(KSlideout)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
