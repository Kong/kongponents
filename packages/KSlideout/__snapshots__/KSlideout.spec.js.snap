// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KSlideout matches snapshot 1`] = `
<div class="k-slideout">
  <!---->
  <!---->
</div>
`;
